### PR TITLE
Various fixes and redesigned logic

### PR DESCRIPTION
**Pull request by NGRichter**
With this pull request I want to incorporate some fixes I made when testing the application.

I also updated the dependencies (except for jline, the new versions did not support some functionality) and everything still seems to be working correctly.

The drop functionality did not work as expected and therefore I remade the logic for that.

The RefLog was also not loaded correctly as it did not request a versions parent but itself, this made it so that no table was based on a table from another version.

I also redesigned the cascading trigger functionality. I tested more than 2 active versions and it did not update version 3 (or 4 for that matter) of a table as the trigger triggering version 2 could not trigger the one going to version 3. (V1 <-> V2 <-> V3 <-> V4 ...)
The way it now works is by exploiting runtime parameters which can be changed locally within a transaction. It now checks if the table that is being written into was already written into in this transaction. This also prevents the endless loop for which pg_trigger_depth() = 0 was used.

I will continue improving and testing QuantumDB. I still have an error with a loop within the RefLog object with a very specific changeset where I change multiple tables twice.

I also have another functionality in mind that returns all tables to their original names with a specific command. This way if someone decides to not use QuantumDB anymore he or she does not have to use the query rewriter anymore.

### CODE REVIEW

File: quantumdb-core/src/main/java/io/quantumdb/core/migration/Migrator.java [hunk 1]
```diff
@@ -64,6 +64,7 @@ public Migrator(Backend backend) {
 	public void migrate(String sourceVersionId, String targetVersionId) throws MigrationException {
 		log.info("Forking from version: {} to version: {}", sourceVersionId, targetVersionId);
 
+		// ToDo give state as parameter as it is already loaded before
```
Comment (michaeldejong) [inline quantumdb-core/src/main/java/io/quantumdb/core/migration/Migrator.java:67]: Still to do? Or no longer needed?

File: quantumdb-postgresql/pom.xml [hunk 2]
```diff
@@ -4,11 +4,24 @@
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
 
 	<modelVersion>4.0.0</modelVersion>
+	<build>
```
Comment (michaeldejong) [inline quantumdb-postgresql/pom.xml:7]: This block isn't needed as the compiler plugin basically ships by default, and uses the properties in the parent POM: https://github.com/quantumdb/quantumdb/blob/master/pom.xml#L22-L23

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java [hunk 3]
```diff
@@ -124,15 +122,61 @@ public void drop(State state, Version version) throws MigrationException {
...
+		// Or first
+		intermediateV = version.getParent();
+		while (intermediateV != null) {
+			if (activeVersions.contains(intermediateV)) {
+				first = intermediateV;
+			}
+			intermediateV = intermediateV.getParent();
+		}
+
+		checkArgument(!(first.equals(version) && last.equals(version)), "Cannot drop the only active version!");
```
Comment (michaeldejong) [inline quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:147]: I could be wrong, but I think it's possible to have multiple active versions that are not directly a child/parent of one another, but instead have an intermediate version that is not active. Examples of this would be a base version, and then having applied a migration using a changeset with >1 operations. Only the base version and the last operation's version would be marked as active.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java [hunk 4]
```diff
@@ -124,15 +122,61 @@ public void drop(State state, Version version) throws MigrationException {
...
+		checkArgument(first.equals(version) || last.equals(version), "The version to drop is not at the beginning or end of the active versions!");
```
Comment (michaeldejong) [inline quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:148]: Is there a reason why this would be a requirement? If there are 3 schemas active at the same time (with or without intermediate version gaps), you should still be able to drop the middle version. This is actually used in some more complex migrations if I remember correctly.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java [hunk 5]
```diff
@@ -112,32 +109,38 @@ public void setColumnsToMigrate(Set<String> columnsToMigrate) {
...
 		return mapping.entrySet().stream()
 				.filter(entry -> entry.getValue().equals(value))
 				.findFirst()
-				.map(Entry::getKey)
-				.get();
+				.map(Entry::getKey);
 	}
 
 	public QueryBuilder createFunctionStatement() {
 		return new QueryBuilder()
```
Comment (michaeldejong) [inline quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java:135]: > I also redesigned the cascading trigger functionality. I tested more than 2 active versions and it did not update version 3 (or 4 for that matter) of a table as the trigger triggering version 2 could not trigger the one going to version 3. (V1 <-> V2 <-> V3 <-> V4 ...)

The intended way for the trigger to function is to only update one table. We want to avoid cascading updates as you described because they're pretty complex and hard to follow / debug.

Instead what should happen with >2 active versions is that you form a fully connected graph using triggers:

```
  V1 <-------> V2 <-------> V3
   ^                        ^
   |                        |
    ------------------------
```

So for a give table active in 3 versions, you'd create sync functions that
- Change table X in V2 when table X in V1 is changed.
- Change table X in V3 when table X in V1 is changed.
- Change table X in V1 when table X in V2 is changed.
- Change table X in V3 when table X in V2 is changed.
- Change table X in V1 when table X in V3 is changed.
- Change table X in V2 when table X in V3 is changed.

> This also prevents the endless loop for which pg_trigger_depth() = 0 was used.

`pg_trigger_depth() = 0` actually prevents triggering other triggers. Not having it will indeed cause endless loops.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/TableDataMigrator.java [hunk 6]
```diff
@@ -231,22 +240,22 @@ private Object parseValue(Type type, String value) {
 			case SMALLINT:
 			case INTEGER:
 				return Integer.parseInt(value);
+			case NUMERIC:
```
Comment (michaeldejong) [inline quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/TableDataMigrator.java:243]: Note that `numeric` can also map to a float or double precision in PG. Those would require a different parser.

**Review by michaeldejong — CHANGES_REQUESTED** [review]
Sorry this review took a while. Busy busy busy. 😅 

Overall very impressed with the work you put in, Nick. This is a pretty difficult codebase to get your head around (understatement), so well done! 

Got some minor things, like left over debug and styling (but one's down to me for not including checkstyle / PMD config files), that need/could be addressed, but minor overall.

Other than that the only things I'm not entirely sure about, are the changes to the triggers (which I think comes down to having different approaches to triggers (not) triggering other triggers), and the active versions changes having to do with the drop command. Perhaps it would be easier to understand/verify if you have a specific changeset / changelog in mind (like a use case), just so I can reproduce and test it. Not sure if you're allowed to share specifics, but example changeset/changelog would be fine too.

So again, well done! And thanks for the contribution!

File: quantumdb-cli/src/main/java/io/quantumdb/cli/Main.java [hunk 7]
```diff
@@ -41,6 +41,7 @@ public static void main(String[] args) throws IOException, SQLException {
 			String command = arguments.remove(0);
 			Command delegate = commands.get(command);
 			if (delegate != null) {
+				writer.write("Executing command: " + command + " with arguments: " + arguments);
```
[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/Main.java:44]: Ah yes, I used the log.debug function a lot, in this class this writer was already available so I used it. I have removed this line in the next commit.

File: quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java [hunk 8]
```diff
@@ -33,7 +33,7 @@ public void perform(CliWriter writer, List<String> arguments) {
 				String user = getArgument(arguments, "username", String.class);
 				String pass = getArgument(arguments, "password", String.class, null);
 
-				config.setUrl("jdbc:postgresql://" + hosts + "/" + catalogName + "?targetServerType=master\"");
+				config.setUrl("jdbc:postgresql://" + hosts); // Postgresql driver 9.3 vs 42: + "/" + catalogName + "?targetServerType=master\""
```
[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java:36]: Due to the dependency change from PostgreSQL JDBC Driver 9.3 to 42 the last part was not needed anymore. I included it at the end should I have to switch back. Everything seems to be working correctly so I have removed the comment.

File: quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java [hunk 9]
```diff
@@ -42,7 +42,7 @@ public AddColumn toOperation() {
 		String defaultExpression = column.getDefaultExpression();
 
 		List<Hint> hints = Lists.newArrayList();
-		if (column.isNullable()) {
+		if (!column.isNullable()) {
```
[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java:45]: This bug happened due to the default value nullable gets if it is not included in the column definition. In the next commit I also changed the default value of nullable:
`column.setNullable(Boolean.TRUE.toString().equals(element.getAttributes().get("nullable")));`
To:
`column.setNullable(!Boolean.FALSE.toString().equals(element.getAttributes().get("nullable")));`
Now when it is not specified, the column is by default nullable instead of not null.

File: quantumdb-postgresql/pom.xml (hunk 2 above)
Comment (NGRichter) [reply quantumdb-postgresql/pom.xml:7]: Ah ok, Maven is still sort of a mystery for me on how to correctly configure it.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 3 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:147]: Hmm, this one is indeed difficult to solve as QuantumDB does not show the user any intermediate versions (only the last one from a changeset). I had a problem where I added 7 foreign key constraint in 1 changeset and when I wanted to drop the version that added it, there were 2 "ghost" tables that were not removed. See the changesets below.

I think that users should keep in mind that 1 changeset should be considered 1 atomic change. If they want to reverse the change, it is either every change in that changeset or none.

I have changed the drop functionality between now and when I submitted this pull request because I noticed an error. 

I will think about what the exact drop functionality should be. If you take a look at the documentation at [https://quantumdb.io/docs/master/#changesets](https://quantumdb.io/docs/master/#changesets), whenever you drop the "oldest" version, in this case 8fb3b4e71a, all the intermediate versions until the next active version (cf3323a79b) should also be dropped (tables will only be dropped if they are not part of the next active version).

If the newest active version gets dropped, also the versions in the changeset to get to the newest version should be dropped, in the example this would be both 72ae28b49a and cc92b7b121. However, if you drop an intermediate active version, what would the functionality be? You use the intermediate versions to get to the next versions, so if you want to drop that one, how can you get to the newest versions? I think the functionality, in this case, should be to only drop the version from the active list and not drop any tables until they are part of either the oldest or newest version to be dropped.

The below changesets can be tested on a database made with [HammerDB](https://www.hammerdb.com/download.html)
```
  <changeset id="add_foreignkeys" author="Nick Richter">
    <description>Add the foreignkey constrains</description>
    <operations>
		<addForeignKey tableName="district" columnNames="d_w_id" referencesTableName="warehouse" referencesColumnNames="w_id" name="district_warehouse_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="customer" columnNames="c_w_id,c_d_id" referencesTableName="district" referencesColumnNames="d_w_id,d_id" name="customer_district_id_fk" onDelete="CASCADE" />
    </operations>
  </changeset>
  
  <changeset id="add_other_foreignkeys" author="Nick Richter">
    <description>Add all other foreignkeys</description>
    <operations>
		<addForeignKey tableName="history" columnNames="h_w_id,h_d_id" referencesTableName="district" referencesColumnNames="d_w_id,d_id" name="history_district_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="orders" columnNames="o_w_id,o_d_id,o_c_id" referencesTableName="customer" referencesColumnNames="c_w_id,c_d_id,c_id" name="orders_customers_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="new_order" columnNames="no_w_id,no_d_id,no_o_id" referencesTableName="orders" referencesColumnNames="o_w_id,o_d_id,o_id" name="neworder_orders_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="stock" columnNames="s_w_id" referencesTableName="warehouse" referencesColumnNames="w_id" name="stock_warehouse_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="stock" columnNames="s_i_id" referencesTableName="item" referencesColumnNames="i_id" name="stock_item_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="order_line" columnNames="ol_w_id,ol_d_id,ol_o_id" referencesTableName="orders" referencesColumnNames="o_w_id,o_d_id,o_id" name="orderline_orders_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="order_line" columnNames="ol_supply_w_id,ol_i_id" referencesTableName="stock" referencesColumnNames="s_w_id,s_i_id" name="orderline_stock_id_fk" onDelete="CASCADE" />
    </operations>

  </changeset>
  <changeset id="add_column" author="Nick Richter">
    <description>Add a new tax column to order_line</description>
    <operations>
		<addColumn tableName="order_line">
        <column name="ol_tax" type="numeric" defaultExpression="0" nullable="false" />
      </addColumn>
    </operations>
  </changeset>
```

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 4 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:148]: Please see my reply above.
I think that you use the intermediate version to get to the next versions, which is why you cannot just drop the middle part. If you have a migration scenario that includes dropping intermediate versions, could you send it to me?

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java (hunk 5 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java:135]: I think it will just be easier to only create the triggers between subsequent versions. If you have to take into account all the additional changes between, for example, V1 and V4 when you could just let the other V2 and V3 triggers deal with their own (smaller) changes, this would decrease the complexity a lot. The way I now implemented it is that is can only propagate in one way and never go back and visit the previous one.

Additionaly, having pg_trigger_depth() = 0 would just completely disable any manually made triggers (they are not supported yet, but if they were this would not work).

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/TableDataMigrator.java (hunk 6 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/TableDataMigrator.java:243]: I have not had any problems with this but I have changed it to the double parser as no information is thrown away:
`return Double.parseDouble(value)`

**Comment by NGRichter** [comment]
I will work on the issues that you commented on.
I have some additional functionality, such as the UNIQUE constraint, that I will also add in the next commit. 
The IDENTITY hint has been renamed to PRIMARY_KEY and the UNIQUE hint has been added.

**Review by michaeldejong — CHANGES_REQUESTED** [review]
Can you split this PR up into smaller ones? It's now +1100 / -700 LoC and it's touching almost every part of the codebase for many different features/changes. This is making things really hard to track/review (especially since some of these changes are probably not the direction we want to go in with QuantumDB).

I'm also not convinced the unique hint is the way to go here. Are you trying to solve some specific issue you can't do with `<createIndex unique="true" ...`?

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 4 above)
Comment (michaeldejong) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:147]: If you have a more complex migration with multiple steps for instance. This is also why the triggers form a fully connected graph and not a single line between versions:

Let's say you start with V1 being active, and are executing a migration that has multiple intermediate versions. This is roughly what the timeline would look like (excuse the terrible ASCII art).

```
Step 0: Baseline

V1


Step 1: Create intermediate version

V1 <----> V2


Step 2: Create intermediate version

V1 <----> V2 <----> V3
 ^                  ^
  \________________/


Step 3: Drop middle intermediate version

V1 <--------------> V3


Step 4: Create intermediate version

V1 <--------------> V3 <----> V4
 ^                            ^
  \__________________________/

Step 5: Drop middle intermediate version

V1 <------------------------> V4

Step 6: Create intermediate version

V1 <------------------------> V4 <----> V5
 ^                                      ^
  \____________________________________/

Step 7: Drop middle intermediate version

V1 <----------------------------------> V5
```

Things get even more complex when you try something like A/B testing different changes (not currently supported with the CLI itself, but the core package does support it). Imagine you're starting with a base version V1, and want to try out two different changes to V1 simultaneously. VA and VB both share a common parent, but they don't rely on each other.

```
   V1
  /  \
VA    VB
```

Because of this complexity QuantumDB constructs the triggers in such a way that a fully connected graph is formed between all (semi-)active versions. This would be pretty hard if you're relying on triggers triggering other triggers.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java (hunk 5 above)
Comment (michaeldejong) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java:134]: > I think it will just be easier to only create the triggers between subsequent versions. If you have to take into account all the additional changes between, for example, V1 and V4 when you could just let the other V2 and V3 triggers deal with their own (smaller) changes, this would decrease the complexity a lot. The way I now implemented it is that is can only propagate in one way and never go back and visit the previous one.

See my previous point why this doesn't work for the more complex migrations.

> Additionaly, having pg_trigger_depth() = 0 would just completely disable any manually made triggers (they are not supported yet, but if they were this would not work).

That's not correct. We're using an "after" trigger to sync the data, which is a trigger that runs after any "before" triggers that want to modify the data before persisting it. You can still safely use other "manually made" triggers. Here's an example with SQL so you can reproduce it locally:

```sql
-- Create base tables A and B
CREATE TABLE a (id bigserial, name text, last_modified timestamptz, primary key (id));
CREATE TABLE b (id bigserial, name text, last_modified timestamptz, primary key (id));


-- Create data triggers for A and B, which update the last_modified column anytime something is changed.
CREATE OR REPLACE FUNCTION set_last_modified() RETURNS TRIGGER AS $$
	DECLARE result record;
	BEGIN
		NEW.last_modified = statement_timestamp();
		RETURN NEW;
	END;
$$ LANGUAGE 'plpgsql';

CREATE TRIGGER t3
BEFORE INSERT OR UPDATE
ON a
FOR EACH ROW
EXECUTE PROCEDURE set_last_modified();

CREATE TRIGGER t4
BEFORE INSERT OR UPDATE
ON b
FOR EACH ROW
EXECUTE PROCEDURE set_last_modified();


-- Create triggers and update functions between tables A and B
CREATE OR REPLACE FUNCTION f1()
RETURNS TRIGGER AS $$
BEGIN
  IF TG_OP = 'INSERT' THEN
    INSERT INTO b (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
  ELSIF TG_OP = 'UPDATE' THEN
    LOOP
      UPDATE b
        SET id = NEW.id, name = NEW.name, last_modified = NEW.last_modified
        WHERE id = OLD.id;
      IF found THEN EXIT; END IF;
      BEGIN
        INSERT INTO b (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
      EXIT;
      EXCEPTION WHEN unique_violation THEN END;
	END LOOP;
  ELSIF TG_OP = 'DELETE' THEN
    DELETE FROM b
      WHERE id = OLD.id;
  END IF;
  RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';


CREATE TRIGGER t1
AFTER INSERT OR UPDATE OR DELETE
ON a
FOR EACH ROW
WHEN (pg_trigger_depth() = 0)
EXECUTE PROCEDURE f1();


CREATE OR REPLACE FUNCTION f2()
RETURNS TRIGGER AS $$
BEGIN
  IF TG_OP = 'INSERT' THEN
    INSERT INTO a (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
  ELSIF TG_OP = 'UPDATE' THEN
    LOOP
      UPDATE a
        SET id = NEW.id, name = NEW.name, last_modified = NEW.last_modified
        WHERE id = OLD.id;
      IF found THEN EXIT; END IF;
      BEGIN
        INSERT INTO a (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
      EXIT;
      EXCEPTION WHEN unique_violation THEN END;
	END LOOP;
  ELSIF TG_OP = 'DELETE' THEN
    DELETE FROM a
      WHERE id = OLD.id;
  END IF;
  RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';


CREATE TRIGGER t2
AFTER INSERT OR UPDATE OR DELETE
ON b
FOR EACH ROW
WHEN (pg_trigger_depth() = 0)
EXECUTE PROCEDURE f2();


-- Example showcasing "manual" triggers. Both records printed will show the updated "last_modified" column.
INSERT INTO a (name) VALUES ('Michael') RETURNING *;
SELECT * FROM b;
```

**Comment by NGRichter** [comment]
I will see what I can do in terms of breaking up this pull request.

On the matter of the unique constraint, I did some more research and from PostgreSQL 10 onward both the unique constraint and unique index have exactly the same functionality. I thought I saw some posts that mentioned that you could not create a foreign key constraint on the unique index while you could on the unique constraint.

It started when I wanted to rename the 'identity' hint to 'primary key' as an identity column is actually an auto incrementing column. I then noticed that the unique constraint was not implemented yet so I started to work on that.

File: quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java [hunk 10]
```diff
@@ -42,17 +42,20 @@ public AddColumn toOperation() {
...
 		if (column.isPrimaryKey()) {
-			hints.add(Hint.IDENTITY);
+			hints.add(Hint.PRIMARY_KEY);
+		}
+		if (column.isUnique()) {
+			hints.add(Hint.UNIQUE);
 		}
 
-		Hint[] hintArray = hints.toArray(new Hint[hints.size()]);
+		Hint[] hintArray = hints.toArray(new Hint[0]);
```
[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java:58]: I use IntelliJ and it says the following:
IntelliJ_toArray
Whenever it says something and it makes sense I just select the 'solution' it suggests.

File: quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java [hunk 11]
```diff
@@ -33,7 +33,7 @@ public void perform(CliWriter writer, List<String> arguments) {
 				String user = getArgument(arguments, "username", String.class);
 				String pass = getArgument(arguments, "password", String.class, null);
 
-				config.setUrl("jdbc:postgresql://" + hosts + "/" + catalogName + "?targetServerType=master\"");
+				config.setUrl("jdbc:postgresql://" + hosts);
```
[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java:36]: I got an error when I updated the driver and solved it by deleting the end part. I have been able to connect to multiple databases using just the "hosts" part.

In the PostgresBackend class the following already existed which made the catalogName redundant in the new driver version. (I actually got the error that I wanted to connect to databasename/databasename)
```
@Override
	@SneakyThrows(ClassNotFoundException.class)
	public Connection connect() throws SQLException {
		Class.forName(driver);
		Connection connection = DriverManager.getConnection(jdbcUrl + "/" + jdbcCatalog, jdbcUser, jdbcPass);
		try (Statement statement = connection.createStatement()) {
			statement.execute("SET SCHEMA 'public';");
		}
		return connection;
	}
```

 I will check if the targetServerType part is still available and how it can be used.

[the remark this answers is not in the document]
Comment (NGRichter) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java:36]: The targetServerType is still used but the value is now changed to 'primary'
https://jdbc.postgresql.org/documentation/head/connect.html

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 3 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: > Just so I can reproduce this:
> 
>     Use HammerDB as the baseline
>     Run quantumdb init to index the database.
>     Add this changelog.xml file.
>     Run quantumdb fork <version for "add_column">
>     Run quantumdb drop <root version>
>     End up with 2 ghost tables that should've been removed.
> 
> Is this correct? And do you happen to know which tables these 2 ghost tables represented?

Yes that is correct, the only thing I did differently was to fork each changeset instead of only the last one.
I will test it again myself to see if it also happens when I only specify the last one.

I have only tested QuantumDB by forking 1 changeset at a time, that is probably the reason why I think about it differently.

Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: You cannot drop the initial version in QuantumDB version 0.4.1-Snapshot as you get a NullPointerException when calling getType() on the root operation (which is null):
```
	@Override
	public void drop(State state, Version version) throws MigrationException {
		// Check that the version's operation is of type DDL, or has no operation (root version).
		checkArgument(version.getOperation() == null || version.getOperation().getType() == Type.DDL);

		RefLog refLog = state.getRefLog();
		Catalog catalog = state.getCatalog();

		List<TableRef> tablesToDrop = refLog.getTableRefs().stream()
				.filter(tableRef -> tableRef.getVersions().contains(version))
				.filter(tableRef -> tableRef.getVersions().stream()
						.filter(otherVersion -> **otherVersion.getOperation().getType()** == Type.DDL)
						.count() == 1)
				.collect(Collectors.toList());
```

Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: Using the "old" drop logic and what you above described it only deleted the 'district' table, the rest of the tables were not dropped when dropping the initial state.

File: quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (michaeldejong) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/xml/XmlAddColumn.java:58]: TIL: Looks like `toArray(new Hint[0])` is now faster in modern JVMs than `toArray(new Hint[hints.size()])` (https://youtrack.jetbrains.com/issue/IDEA-180847). So 👍

File: quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (michaeldejong) [reply quantumdb-cli/src/main/java/io/quantumdb/cli/commands/Init.java:36]: Ah that would explain the issues indeed.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 3 above)
Comment (michaeldejong) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: I'll double check myself as well, as you should be able to drop the initial version after the first fork. Will let you know what I find.

**Comment by michaeldejong** [comment]
> On the matter of the unique constraint, I did some more research and from PostgreSQL 10 onward both the unique constraint and unique index have exactly the same functionality. I thought I saw some posts that mentioned that you could not create a foreign key constraint on the unique index while you could on the unique constraint.

Interesting. That's the first time I heard about it. Please let me know if you have a link for me to read up on.

> It started when I wanted to rename the 'identity' hint to 'primary key' as an identity column is actually an auto incrementing column. I then noticed that the unique constraint was not implemented yet so I started to work on that.

I'm fine with that name change. Primary key / auto increment are closer to terms used by SQL / PG than identity.

File: quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java (hunk 3 above)
Comment (NGRichter) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: This one is with my new drop logic which is correct:
2021-05-12 17_11_09-pgAdmin 4 — Mozilla Firefox

Comment (michaeldejong) [reply quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java:146]: Managed to set up a database with the HammerDB schema yesterday, and ran into the `numeric` issue when trying to init quantumdb. I'll use some of the fixes from this PR to check what's going wrong. I'll keep you updated.

**Comment by NGRichter** [comment]
I have already made 2 other pull requests.
I will close this one, any additional discussion can happen in the other (to be made) pull requests.
